Fix first instance rule being used as rule description for all violations of that rule and other SARIF improvements

Nettozx opened the pull request:
* Fixed issue where the first instance of a rule violation short description would get used for all subsequent rule violations, found that if you make all the rule name and descriptions empty strings, github will default to the instance descriptions
* Added setting for problem.severity
* Changed defaultConfiguration.level to use problem.severity instead of security-severity
* Added more levels for security-severity
* Added reporting of cwe ID
* Fixed issue with uncrustify version detection
* Added unit tests for sarif output

Before:
![cppcheck_original_problem](https://github.com/user-attachments/assets/badb787a-0f2c-4809-a41f-a1d1b66b39d3)

After:
<img width="925" height="1636" alt="image" src="https://github.com/user-attachments/assets/d40f5f9f-186c-4367-9290-9d73cbd23d55" />

firewave commented on cli/cppcheckexecutor.cpp:
```diff
@@ -199,12 +228,252 @@ namespace {
+            };
+
+            // Mapping of Cppcheck rule IDs to descriptions
+            static const RuleDescription ruleDescriptions[] = {
+                {"nullPointer", "Null pointer dereference", 
+                 "Detects when code dereferences a pointer that may be null, which can lead to undefined behavior or program crashes."},
```
As this description possibly exists in other place this should probably generated so it does how have to maintained in multiple places and is not missed if new ones are added.

Same for the code below.

This could probably be integrated with #7439 if that information would be externally available.

Maybe this could be generated at run-time ny leveraging the `getErrorMessages()` function by collecting the `ErrorMessages` instead of generating some output. Possibly adding some like `ErrorID` which contains the information to generate this - similar to/instead of `CWE`.

firewave commented:
Thanks for you contribution.

I added a remark on how to keep this in sync for future changes. Possibly not something which should addressed before/in this PR.

Also something like `--errorlist-sarif` might make sense.

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -94,31 +102,52 @@ namespace {
+                    if (isSecurityRelatedFinding(finding.id)) {
+                        double securitySeverity = 0;
+                        if (ErrorLogger::isCriticalErrorId(finding.id)) {
+                            securitySeverity = 9.9; // critical = 9.0+
```
I am not sure.. a "critical id" means that cppcheck analysis is aborted so probably the cppcheck configuration is not ideal. 

It does not indicate at all if there is a security issue or not.

Is there some documentation that says how the security severity should be assigned in this case?

Nettozx commented on cli/cppcheckexecutor.cpp:
```diff
@@ -94,31 +102,52 @@ namespace {
+                            securitySeverity = 9.9; // critical = 9.0+
+                        }
+                        else if (finding.severity == Severity::error) {
+                            securitySeverity = 8.5; // high = 7.0 to 8.9
```
You are correct, I had a misconception about isCriticalErrorId()

Nettozx commented:
@firewave  @danmar I have implemented a generic approach that will not require ruleID maintenance, but like @firewave  said the security related ruleID list should get moved to --errorlist-sarif or something in the future

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -78,71 +83,379 @@
+            result = std::regex_replace(result, std::regex(R"(assert\([^)]+\))"), "assert(condition)");
+            result = std::regex_replace(result, std::regex(R"(for expression '[^']*')"), "for expression");
+
+            // === GENERIC VARIABLE/IDENTIFIER REPLACEMENT ===
```
some of these generic replacements are probably fine to keep

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -78,71 +83,379 @@
+            {
+                const std::string& description = fullDescription ? it->second.second : it->second.first;
+                // Convert instance-specific descriptions to generic ones
+                return makeGeneric(description);
```
result should be cached. or can a ruleId only be requested once?

Nettozx replied:
yes rule's are only needed once

danmar replied:
please clarify.. so there will only be 1 request for each ruleId? So a cache is redundant?

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -78,71 +83,379 @@
+                "Format specifier in format string requires different argument type than provided.");
+            result = std::regex_replace(
+                result,
+                std::regex(R"(%\w+ in format string \(no\. \d+\) requires '[^']*' but the argument type is [^.]*\.)"),
```
I don't have a good feeling about so many regexes.

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -78,71 +83,379 @@
+            result = std::regex_replace(
+                result,
+                std::regex(
+                    R"(%[a-zA-Z]+ in format string \(no\. \d+\) requires '[^']*' but the argument type is '[^']*'\.)"),
```
I am not sure if `--errorlist` messages need to be as specific as they are. It's possible that the more generic messages would be better.

Nettozx replied:
got it, i'll come back to this when I have some free time

danmar commented:
In the future.. if I don't respond for couple of days.. feel free to ping me.
It happens that I loose track of PRs.

Nettozx commented:
@danmar  I have simplified the approach significantly. I added genericMessage output for ErrorMessage class like you recommended and have reduced the regex to a minimum. I also removed the security violation list of errors and now I just add security tag if a CWE ID exists.

Nettozx commented:
Actually I just made a discovery. If I make all the descriptions for the rules blank strings, github will default to the instance specific description for each and this solves the problem. So this should greatly simplify the addition.

Nettozx commented:
@danmar  @firewave  I updated the description of this PR with latest results and changes. This is ready for review now.

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -78,71 +78,117 @@
+                    // this makes it so not all the violations have the same description.
                     picojson::object rule;
-                    rule["id"] = picojson::value(finding.id);
+                    rule["id"]   = picojson::value(finding.id);
```
this extra space is redundant

Nettozx commented on test/testerrorlogger.cpp:
```diff
@@ -315,7 +315,7 @@ class TestErrorLogger : public TestFixture {
 
     #define testReportType(reportType, severity, errorId, expectedClassification, expectedGuideline) \
-        testReportType_(__FILE__, __LINE__, reportType, severity, errorId, expectedClassification, expectedGuideline)
+    testReportType_(__FILE__, __LINE__, reportType, severity, errorId, expectedClassification, expectedGuideline)
```
some of these style changes were due to running uncrustify based on your repo's rules, but I can revert them and not run it

Nettozx commented:
> > Actually I just made a discovery. If I make all the descriptions for the rules blank strings, github will default to the instance specific description for each and this solves the problem. So this should greatly simplify the addition.
> 
> That is so much better 👍
> 
> Sounds like it will both simplify our job and make the user experience better.

@danmar  this is ready for review again

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -92,33 +92,65 @@ namespace {
+                        }
+                        if (securitySeverity > 0.0) {
+                            std::ostringstream ss;
+                            ss << securitySeverity;
```
this still does not use fixed number of decimals so the output will vary..

danmar replied:
imho I would prefer to only write 1 decimal for 9.9 etc.

Nettozx commented:
> CI: I think you can fix the "dmake" failures by just running `make run-dmake` in the cppcheck project folder.

@danmar  I fixed the dmake issue.

Nettozx commented:
@danmar this is ready for review

Nettozx commented:
> @Nettozx there are 5 failing tests in the CI have you looked at those?

@danmar its all passing now

danmar commented on test/testsarif.cpp:
```diff
@@ -0,0 +1,1111 @@
+
+    // Shared test code with various error types
+    const std::string testCode = R"(
+#include <cstdio>
```
I feel such tests should be written in test/cli/test_sarifreport.py instead

danmar replied:
a proper unit test would create a SarifReport. Add hardcoded findings to it. then check the state of the SarifReport in some way.

If we want to check the json output I would suggest that each `SarifReport::serialize*()` is tested separately.

danmar commented:
> I will get back to this eventually when I have more time to donate

sure! in the meantime maybe somebody else wants to finish this PR.. having proper SARIF output is something I do think is very important.

Nettozx commented:
@danmar This should be ready for review again now

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -710,19 +561,21 @@ void StdLogger::reportErr(const ErrorMessage &msg)
-    if (!mSettings.emitDuplicates && !mShownErrors.insert(msgStr).second)
-        return;
+        // Alert only about unique errors
+        if (!mSettings.emitDuplicates && !mShownErrors.insert(msgStr).second)
```
It seems to me that with this code duplicates will be written in the sarif output that seems unfortunate.

Nettozx replied:
This code does not trigger for mSettings.outputFormat == Settings::OutputFormat::sarif, its in the else block
